lib/TPP/Transforms/LinalgTiling.cpp

```diff
@@ -0,0 +1,237 @@
+
+//===- LinalgTiling.cpp -----------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements parallel loop insertion for tiling.
+//
+//===----------------------------------------------------------------------===//
+#include "TPP/IR/TilingUtils.h"
+#include "TPP/Transforms/Transforms.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/IR/AffineValueMap.h"
+#include "mlir/Dialect/Affine/Utils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/IntegerSet.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "llvm/Support/Debug.h"
+#include <algorithm>
+#include <iostream>
+#include <list>
+#include <set>
+#define DEBUG_TYPE "linalg-tiling"
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DECL_LINALGTILING
+#define GEN_PASS_DEF_LINALGTILING
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::tpp;
+using namespace std;
+
+namespace mlir {
+namespace tpp {
+struct LinalgOpTiling : OpRewritePattern<linalg::BatchReduceMatmulOp> {
+  using OpRewritePattern<linalg::BatchReduceMatmulOp>::OpRewritePattern;
+
+  LinalgOpTiling(MLIRContext *ctx, LinalgTilingOptions tilingoptions)
+      : OpRewritePattern(ctx), options(tilingoptions) {}
+
+  LogicalResult matchAndRewrite(linalg::BatchReduceMatmulOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    //  Get the MXN tile shape from the user input
+    std::vector<int64_t> tileShapeM(options.mTileShape.begin(),
+                                    options.mTileShape.end());
+    std::vector<int64_t> tileShapeN(options.nTileShape.begin(),
+                                    options.nTileShape.end());
+    std::vector<int64_t> finaltile(3);
+    std::vector<int64_t> resulttile(2);
+    if (tileShapeM.size() == 2 && tileShapeN.size() == 2) {
+      if (tileShapeM[1] == tileShapeN[0]) {
+        finaltile[0] = tileShapeM[0];
+        finaltile[1] = tileShapeN[1];
+        finaltile[2] = tileShapeM[1];
+        resulttile[0] = tileShapeM[0];
+        resulttile[1] = tileShapeN[1];
+      } else {
+        return failure();
+      }
+    } else {
+      return failure();
+    }
```

Comment:
Could you refactor this into guard clauses?

Reply:
Yes. Refactored the code into a guard clause.